host_core/lib/host_core/vhost/heartbeats.ex

```diff
@@ -4,26 +4,53 @@ defmodule HostCore.Vhost.Heartbeats do
   virtual host from which the heartbeat eminates.
   """
 
+  @thirty_seconds 30_000
+
+  use GenServer, restart: :transient
+
   alias HostCore.Actors.ActorSupervisor
   alias HostCore.CloudEvent
   alias HostCore.Providers.ProviderSupervisor
   alias Timex.Format.Duration.Formatters.Humanized
 
+  def start_link(host_pid, host_key) when is_pid(host_pid) do
+    GenServer.start_link(__MODULE__, {host_pid, host_key}, name: String.to_atom("hb-#{host_key}"))
+  end
+
+  @impl true
+  def init({host_pid, host_key}) do
+    :timer.send_interval(@thirty_seconds, self(), :publish_heartbeat)
+
+    {:ok, {host_pid, host_key}}
+  end
+
+  @impl true
+  def handle_info(:publish_heartbeat, state) do
+    publish_heartbeat(state)
+    {:noreply, state}
+  end
+
+  defp publish_heartbeat({host_pid, _host_key}) do
+    if Process.alive?(host_pid) do
+      state = HostCore.Vhost.VirtualHost.full_state(host_pid)
+
+      state
+      |> generate_heartbeat()
+      |> CloudEvent.publish(state.config.lattice_prefix)
+    end
+  end
+
   def generate_heartbeat(state) do
     config = state.config
-    # TODO: for large numbers of actors this heartbeat becomes prohibitively large and expensive
-    # refactor to only emit instance count
+
     actors =
       config.host_key
       |> ActorSupervisor.all_actors_for_hb()
-      |> Enum.map(fn {k, iid} -> %{public_key: k, instance_id: iid} end)
+      |> Map.new()
```

Comment:
Is this needed? If so, is it not needed for `providers`?

Reply:
providers is a list, this is a map. yes it's needed.

Reply:
It's not clear why `all_actors_for_hb` is a list that needs to be turned into a map, whereas `all_providers_for_hb` is a list that doesn't need to be turned into a map. Why not make `all_actors_for_hb` include the call to `Map.new`? And consider adding type specs to these functions to help readers understand what these structures are intended to be